### PR TITLE
Change Tier Currency to Host Currency on Collective Approval

### PR DESCRIPTION
**Pull request by SudharakaP**
Resolve https://github.com/opencollective/opencollective/issues/7015

[tier-change.webm](https://github.com/opencollective/opencollective-api/assets/12435965/96645303-2435-4dfa-b6ba-330357716892)

### CODE REVIEW

File: server/graphql/v2/mutation/HostApplicationMutations.ts [hunk 1]
```diff
@@ -305,6 +305,12 @@ const approveApplication = async (host, collective, req) => {
       transaction,
     });
 
+    // Convert all active tiers to host currency
+    const tiers = await models.Tier.findAll({ where: { CollectiveId: collective.id, deletedAt: null }, transaction });
+    tiers.map(tier => {
+      tier.update({ currency: host.currency }, { transaction });
+    });
+
```
Comment (Betree) [inline server/graphql/v2/mutation/HostApplicationMutations.ts:313]: A simpler/more performant alternative to that:

```suggestion
    await models.Tier.update({ currency: host.currency }, { where: { CollectiveId: collective.id }, transaction });

```

**Review by Betree — COMMENTED** [review]
It looks like we're already supposed to do that in https://github.com/opencollective/opencollective-api/blob/583b63ab30999bbcbca3c49afb2623c744cb761c/server/models/Collective.ts#L2224. `addHost` seems like the right place to do it, as it will do the change no matter which path we use. We should try to understand why it hasn't worked with the case you're referring to.

File: server/graphql/v2/mutation/HostApplicationMutations.ts (hunk 1 above)
Comment (SudharakaP) [reply server/graphql/v2/mutation/HostApplicationMutations.ts]: Thanks. Done.

**Comment by SudharakaP** [comment]
> It looks like we're already supposed to do that in
> 
> https://github.com/opencollective/opencollective-api/blob/583b63ab30999bbcbca3c49afb2623c744cb761c/server/models/Collective.ts#L2224
> . `addHost` seems like the right place to do it, as it will do the change no matter which path we use. We should try to understand why it hasn't worked with the case you're referring to.

@Betree : This `addHost` is only called when the collective applies for being hosted. This is why in the screen-cast, after applying to be hosted, I have to go again to Tiers (which is now in the correct currency) and 

> It looks like we're already supposed to do that in
> 
> https://github.com/opencollective/opencollective-api/blob/583b63ab30999bbcbca3c49afb2623c744cb761c/server/models/Collective.ts#L2224
> . `addHost` seems like the right place to do it, as it will do the change no matter which path we use. We should try to understand why it hasn't worked with the case you're referring to.

The reason is that the function `addHost` is called through `changeHost`. And `changeHost` is only called when we apply for a host; such as in, 

https://github.com/opencollective/opencollective-api/blob/d8e64744ad0f281bb9377f1b4b594074d9fd34ee/server/graphql/v2/mutation/HostApplicationMutations.ts#L145

or when removing the host, 

https://github.com/opencollective/opencollective-api/blob/d8e64744ad0f281bb9377f1b4b594074d9fd34ee/server/graphql/v2/mutation/HostApplicationMutations.ts#L253

or rejecting the host,

https://github.com/opencollective/opencollective-api/blob/d8e64744ad0f281bb9377f1b4b594074d9fd34ee/server/graphql/v2/mutation/HostApplicationMutations.ts#L353

But it's not called when approving the host;

https://github.com/opencollective/opencollective-api/blob/d8e64744ad0f281bb9377f1b4b594074d9fd34ee/server/graphql/v2/mutation/HostApplicationMutations.ts#L274

**Review by Betree — COMMENTED** [review]
That makes sense, I'm good with this PR then (apart from the small feedback I've shared above).

I've documented https://github.com/opencollective/opencollective/issues/7019 for us to rethink our approach to this.

File: server/graphql/v2/mutation/HostApplicationMutations.ts [hunk 2]
```diff
@@ -305,6 +305,12 @@ const approveApplication = async (host, collective, req) => {
       transaction,
     });
 
+    // Convert all active tiers to host currency
+    await models.Tier.update(
+      { currency: host.currency },
+      { where: { CollectiveId: collective.id }, validate: false, transaction },
```
[the remark this answers is not in the document]
Comment (SudharakaP) [reply server/graphql/v2/mutation/HostApplicationMutations.ts:311]: This won't work as we don't have `ParentCollectiveId` in `Tiers`, so I've changed the code to first get all the child collectives and then update all the tiers relevant to each one of them.

**Review by Betree — APPROVED** [review]
I pushed an improvement to:
- Select only the `id` from children
- Update all tiers in a single SQL query rather that `nbChildren` queries
- Not update tiers that already have host's currency
- Add tests

**Pull request merged**